CIP-array-builtin/README.md

```diff
@@ -0,0 +1,296 @@
+
+2. Failed lookups return a default value provided by the caller:
+`indexArray :: forall a . Integer -> a -> Array a -> a`.
+    - This solution is problematic whenever there is no sensible default value and
+    the user wants the function to fail. Since Plutus Core is strict, it is not possible
+    to pass `error` as the default value without it getting evaluated before the call and
+    terminating execution immediately.
+
+3. Include `lengthOfArray` and require the user to perform appropriate length guards before
+calling `indexArray`.
+    - This is a familiar option from other languages.
+    - By definition arrays are fixed-size and their length is available in constant time.
+    - It also allows users to omit bounds checks when they know a priori that the
+    index is in bounds.
+
+After considering these three options, we concluded that the addition of
+the `lengthOfArray` builtin function is both necessary and sufficient for introducing a
+well-defined array type into the builtin language.
+
+### Constructing arrays
+
+The last required functionality for having a practical interface is the ability to
+construct arrays.
+
+Due to our decision of providing immutable arrays, it is difficult (or very expensive) to build up arrays incrementally.
+A naive approach would require repeated copying and potentially the usage of quadratic
+space and time.
+
+A more appropriate approach would be to construct the array in bulk,
+however that would require an intermediate representation of a collection of elements. Fortunately this already exists in the builtin language in the form of builtin lists.
+We can then naturally introduce a function which transforms lists into arrays: `listToArray`.
+
+Since the only current method of constructing arrays is via `listToArray`, one may ask
+whether the `lengthOfArray` builtin is at all necessary.
+We believe that, indeed, `lengthOfArray` is redundant in the current specification. The reason is
+that traversing the list to construct the array is enough to also compute its length. 
+However, we have decided that `lengthOfArray` should be kept as future work will depend upon its
+existence. Some components of Plutus Core could be natively represented as arrays,
+and there would be no other way to access their length in constant time other than
+through a builtin.
+The [Arrays in `Data`](#arrays-in-data) section considers such a case.
+
+### Slices
+
+Many array-like data structures support cheap _slices_, e.g. a function with the following
+type signature: `slice :: Integer -> Integer -> Array a -> Array a`, which produces a _view_
+of the original array between the two indices (similarly, the same can be achieved using an
+`indexArray` and a `lengthOfArray`).
+
+We do not propose to add `slice` to our builtin set, since it is very easy to build a data
+structure that supports slices on top of `array`, simply by tracking some additional
+integers to track the subset of the array that is in view.
```

Comment:
Except if we do it as a builtin it'll be much much cheaper compared to implementing that logic in Plutus. Especially given that even just having that logic in the program is already costly in terms of size and that is our scarcest resource, unlike builtins real estate, which we can in theory expand a lot.

I don't mind not adding `sliceArray` right away, but let's make the underlying type `Vector` (which has fast slicing) anyway, since we'll probably end up adding `sliceArray` eventually.

Reply:
We can and should benchmark this. I wouldn't expect the overhead in PLC to be that high, since the slicing logic is very simple? But maybe it will be significant.

Reply:
Well, pattern matching builtins [are](https://github.com/IntersectMBO/plutus/pull/5486#issuecomment-2324804683) 20-30% more efficient than encoding pattern matching using `chooseList`/`chooseData` , so the same logic gotta apply here. Not sure if 20-30% is considered "that high" or not though, but then array slicing is probably gonna have even more overhead. `vector` uses this within a call to [`slice`](https://hackage.haskell.org/package/vector-0.13.1.0/docs/src/Data.Vector.Generic.html#slice):

```haskell
checkSlice :: HasCallStack => Checks -> Int -> Int -> Int -> a -> a
{-# INLINE checkSlice #-}
checkSlice kind i m n x
  = check kind (checkSlice_msg i m n) (i >= 0 && m >= 0 && m <= n - i) x
```

and that's 6 builtin calls right there.

Reply:
Then I'll add `slice` to the CIP unless anyone objects

Reply:
> Then I'll add `slice` to the CIP unless anyone objects

Have you done that? Not sure if something's changed or not.

I don't mind not including `sliceArray` right away, however I really want the underlying array representation to be `sliceArray`-friendly, i.e. I want `Vector`.